src/thor/costmatrix.cc

```diff
@@ -326,22 +331,40 @@ void CostMatrix::Initialize(
                     return limits.max_up_transitions == kUnlimitedTransitions;
                   });
 
-  // Add initial sources status
-  for (const auto exp_dir : {MATRIX_FORW, MATRIX_REV}) {
-    const auto count = locs_count_[exp_dir];
-    locs_status_[exp_dir].reserve(count);
-    hierarchy_limits_[exp_dir].resize(count);
-    adjacency_[exp_dir].resize(count);
-    edgestatus_[exp_dir].resize(count);
-    edgelabel_[exp_dir].resize(count);
+  const uint32_t bucketsize = costing_->UnitSize();
+  const float range = kBucketCount * bucketsize;
+
+  // Add initial sources & targets properties
+  for (const auto is_fwd : {MATRIX_FORW, MATRIX_REV}) {
+    const auto count = locs_count_[is_fwd];
+    const auto other_count = locs_count_[!is_fwd];
+
+    const auto& locations = is_fwd ? source_locations : target_locations;
+    const auto& other_locations = is_fwd ? target_locations : source_locations;
+
+    locs_status_[is_fwd].reserve(count);
+    hierarchy_limits_[is_fwd].resize(count);
+    adjacency_[is_fwd].resize(count);
+    edgestatus_[is_fwd].resize(count);
+    edgelabel_[is_fwd].resize(count);
     for (uint32_t i = 0; i < count; i++) {
       // Allocate the adjacency list and hierarchy limits for this source.
       // Use the cost threshold to size the adjacency list.
-      edgelabel_[exp_dir][i].reserve(max_reserved_labels_count_);
-      adjacency_[exp_dir][i].reuse(0, current_cost_threshold_, costing_->UnitSize(),
-                                   &edgelabel_[exp_dir][i]);
-      locs_status_[exp_dir].emplace_back(kMaxThreshold);
-      hierarchy_limits_[exp_dir][i] = hlimits;
+      edgelabel_[is_fwd][i].reserve(max_reserved_labels_count_);
+      locs_status_[is_fwd].emplace_back(kMaxThreshold);
+      hierarchy_limits_[is_fwd][i] = hlimits;
+      // for each source/target init the other direction's astar heuristic
+      auto& ll = locations[i].ll();
+      astar_heuristics_[!is_fwd][i].Init({ll.lng(), ll.lat()}, costing_->AStarCostFactor());
```

Comment:
so a source create a heuristic for the reverse tree, so that each edge on the reverse search can find the clostest/lowest-cost source for the sortcost